rclcpp_action/include/rclcpp_action/client.hpp

```diff
@@ -387,6 +379,16 @@ class Client : public ClientBase
             return;
           }
         }
+
+        {
+          std::lock_guard<std::mutex> guard(goal_handles_mutex_);
+          goal_handles_[goal_handle->get_goal_id()] = goal_handle;
+        }
+
+        promise->set_value(goal_handle);
+        if (options.goal_response_callback) {
+          options.goal_response_callback(future);
```

Comment:
It was moved in response to https://github.com/ros2/rclcpp/pull/701#pullrequestreview-233076772; if  `make_result_aware` is called first, we risk having the result callback triggered before the goal response callback (unlikely, but I *think* this is possible if the action server is running in the same process).

Reply:
Hmm, reflecting on this, it seems to me that the goal response callback should simply be taking a concrete  goal handle instead of a future one. Additionally, failing to make the goal handle result aware should not result in caller code losing access to the goal handle. Getting the future result should throw, but the exception could carry the goal handle (which is still valid).

Reply:
@hidmic I agree with everything you've said. I think it's okay to not backport changes (though maybe we could by supporting two different signatures for the goal response callback).

Reply:
Adjusted the goal_response_callback signature, but I'm still not decided about what to do exception and goal handle bit.

Reply:
I think we can remove the try-catch altogether. I'll open a PR in a moment with a proposed alternative.

Reply:
See https://github.com/ros2/rclcpp/pull/1311.